### ISSUE TITLE
Trying to follow example, Error:  Attribute was supposed to be a Array, but was a String

### ISSUE THREAD

**Issue by drc37** [skryl/qbwc#56]
I am getting an error like this (I do have 2 users in my test account):
```
Attribute was supposed to be a Array, but was a String. -- "--- []\n":
```

Gem:
```
gem "qbwc", github: 'skryl/qbwc'
```

My worker file:
```
class CustomerWorker < QBWC::Worker
  def requests(job)
    puts "Requests Call"
    {
      :customer_query_rq => {:active_status => 'All', :max_returned => '1'}
    }
  end

  def should_run?(job)
    true
  end

  def handle_response(r, session, job, request, data)
    # handle_response will get customers in groups of 100. When this is 0, we're done.
    complete = r['xml_attributes']['iteratorRemainingCount'] == '0'

    r['customer_ret'].each do |qb_cus|
      qb_id = qb_cus['list_id']
      qb_name = qb_cus['name']
      ap "#{qb_id} #{qb_name}"
    end
  end
end
```

Error output:
```
ActiveRecord::SerializationTypeMismatch - Attribute was supposed to be a Array, but was a String. -- "--- []\n":
  activerecord (4.2.0) lib/active_record/coders/yaml_column.rb:17:in `dump'
  activerecord (4.2.0) lib/active_record/type/serialized.rb:26:in `type_cast_for_database'
  activerecord (4.2.0) lib/active_record/connection_adapters/abstract/quoting.rb:13:in `quote'
  activerecord (4.2.0) lib/active_record/sanitization.rb:167:in `quote_bound_value'
  activerecord (4.2.0) lib/active_record/sanitization.rb:109:in `block in sanitize_sql_hash_for_assignment'
  activerecord (4.2.0) lib/active_record/sanitization.rb:108:in `sanitize_sql_hash_for_assignment'
  activerecord (4.2.0) lib/active_record/sanitization.rb:40:in `sanitize_sql_for_assignment'
  activerecord (4.2.0) lib/active_record/relation.rb:332:in `update_all'
   () Users/david/.rvm/gems/ruby-2.2.0/bundler/gems/qbwc-066129549c1a/lib/qbwc/active_record/job.rb:64:in `requests='
   () Users/david/.rvm/gems/ruby-2.2.0/bundler/gems/qbwc-066129549c1a/lib/qbwc/job.rb:107:in `reset'
   () Users/david/.rvm/gems/ruby-2.2.0/bundler/gems/qbwc-066129549c1a/lib/qbwc/session.rb:98:in `reset'
   () Users/david/.rvm/gems/ruby-2.2.0/bundler/gems/qbwc-066129549c1a/lib/qbwc/session.rb:24:in `initialize'
   () Users/david/.rvm/gems/ruby-2.2.0/bundler/gems/qbwc-066129549c1a/lib/qbwc/active_record/session.rb:20:in `initialize'
   () Users/david/.rvm/gems/ruby-2.2.0/bundler/gems/qbwc-066129549c1a/lib/qbwc/controller.rb:123:in `authenticate'
  actionpack (4.2.0) lib/action_controller/metal/implicit_render.rb:4:in `send_action'
  actionpack (4.2.0) lib/abstract_controller/base.rb:198:in `process_action'
  actionpack (4.2.0) lib/action_controller/metal/rendering.rb:10:in `process_action'
  actionpack (4.2.0) lib/abstract_controller/callbacks.rb:20:in `block in process_action'
  activesupport (4.2.0) lib/active_support/callbacks.rb:117:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:219:in `block in halting_and_conditional'
  activesupport (4.2.0) lib/active_support/callbacks.rb:151:in `block in halting_and_conditional'
  activesupport (4.2.0) lib/active_support/callbacks.rb:151:in `block in halting_and_conditional'
  activesupport (4.2.0) lib/active_support/callbacks.rb:151:in `block in halting_and_conditional'
  activesupport (4.2.0) lib/active_support/callbacks.rb:308:in `block (2 levels) in halting'
  wash_out (0.9.2) lib/wash_out/dispatcher.rb:139:in `_catch_soap_errors'
  activesupport (4.2.0) lib/active_support/callbacks.rb:427:in `block in make_lambda'
```

**Comment by yos314** [comment]
I am also getting this error (And was going to be writing what I had found about it later today once I get back to the machine that I am coding on).  

The -- "--- []\n" is somehow getting put into QBWC::ActiveRecord::Job.data column for the Job Record.

I believe it has to do with something in 4.2 that doesn't happen in =<4.1.  

I had forked the repository yesterday to see what the problem could be, and running rake:test came up with something like 46 failures and 2 errors.  I had used Ruby 2.2 and a fresh rvm gemset for the project.  I'll post more later this evening, just wanted to state that I had also hit this same problem.

**Comment by drc37** [comment]
Yeah, I downgraded to rails 4.1.9 and everything seemed to work after that.

**Comment by JasonBarnabe** [comment]
Well, the relevant code would be https://github.com/skryl/qbwc/blob/master/lib/qbwc/active_record/job.rb#L4. Note that @drc37's problem is about the requests column and @yos314 is talking about the data column.

I'm not sure what could be different in 4.2. Maybe it serializes things differently? What happens if you clear out the jobs, start up with 4.2, then add your job?

**Comment by JasonBarnabe** [comment]
Maybe https://github.com/rails/rails/issues/18037

**Comment by yos314** [comment]
Serialization seems to be automatic now, so getting rid of the serialize statements inside the active_record/jobs.rb fixes that error.  From a clean install of the gem and running rake test, the following pops up. [Gist](https://gist.github.com/yos314/ac7b1a5375ada54ab250)

I've been tinkering around with it in my rails4.2 branch.  I managed to get rid of the serialization issue in the jobs.rb by adding a version check.  But there are a pile of errors after that which I'm trying to sort through.

**Comment by JasonBarnabe** [comment]
Kind of surprising so much would be different with Rails 4.2 as I've been using it with 4.0 and 4.1 with no problems. PR would be welcome once you've gotten it figured out, because I'll have to upgrade my stuff eventually...

**Comment by JasonBarnabe** [comment]
Now that I actually look at the code and https://github.com/rails/rails/issues/18037, this doesn't seem right:

https://github.com/skryl/qbwc/blob/master/lib/qbwc/active_record/job.rb#L64
https://github.com/skryl/qbwc/blob/master/lib/qbwc/active_record/job.rb#L82

The `serialize` stuff should handle this without calling `to_yaml`. I can confirm that with 4.1, removing the two `to_yaml`s while leaving the `serialize` doesn't cause any tests to fail. Can you confirm with your 4.2?

I think just taking out the `serialize`s won't work, because then you'll be getting strings out of the DB when reading instead of the arrays or whatnot, wouldn't you?

**Comment by yos314** [comment]
I can try that. 

 I also am going to try getting rails 4.1 set up in another rvm/gemset to be able to see how things react there too.

**Comment by yos314** [comment]
Ok.  I believe that changing the to_yaml is exactly what is required.  [Here is the gist](https://gist.github.com/yos314/8ab1acdc28314cdcee5f).  There are a lot of deprecation errors and looks like some testing behaviour changed.  I'll see if I can clean up what is left in that gist too.

**Comment by yos314** [comment]
I just tested my soon-to-be PR in my application, and amazingly, it worked!  I have to say that it is the first time in my 2 or so years of attempting to get qbwc to work off and on to actually get a connection to spin off properly!

**Comment by yos314** [comment]
Oh, you got it.  I wasn't sure if the to_yaml was required for rails 4.1 so I had made a PR that checked for versioning.  If that works for 4.1 also, then that commit you just did will solve all the problems.

**Comment by JasonBarnabe** [comment]
Yeah, me and @rchekaluk have been working on it since December, and it's much better than before.

I don't that my commit solved *all* the problems - you'd still see the warnings and fails running `rake test` that you posted, but it should have made it usable.

**Comment by yos314** [comment]
Very true.  I am hoping to help a bit when I find stuff.  I do know that I did have a problem in previous versions because my application really needs to use multiple <IncludeRetElement >STRTYPE</IncludeRetElement>'s and a normal hash wouldn't allow more than one.

**Comment by JasonBarnabe** [comment]
@drc37 is this working for you now?

**Comment by drc37** [comment]
Now I feel bad, I have only been working on it for six hours and get it working. Well you guys have been working on it for months.

I was right out this afternoon to make sure it works with 4.2

**Comment by drc37** [comment]
It wasn't working form on 4.2.0 this morning. It just keeps sending error messages with no progress.

![screen shot 2015-02-03 at 8 49 45 am](https://cloud.githubusercontent.com/assets/162541/6023411/e81fd1ae-ab81-11e4-9c04-7ecd696e3160.png)

**Comment by JasonBarnabe** [comment]
The order of your elements is important. Change
```ruby
:customer_query_rq => {:active_status => 'All', :max_returned => '1'}
```
to
```ruby
:customer_query_rq => {:max_returned => '1', :active_status => 'All'}
```
See also https://github.com/skryl/qbwc/issues/53

**Comment by drc37** [comment]
Ok, reordering seems to fix the issue with 4.2.0. Again, thanks for the help.
